Reading FITS function added to sji.py

BaptistePellorceAstro opened the pull request:
As a first part of SJICube Refactor, I created a reading FITS function named `read_iris_sji_level2_fits`. For now, the function is independent and return an NDCube instance that will be change to SJICube when it will be refactored. 

Most of this new function is clearly inspired by spectograph's FITS reading function and sji's initialization.

Cadair commented on irispy/sji.py:
```diff
@@ -878,3 +879,73 @@ def dustbuster(mc):
+    """
+
+    # Open a fits file
+    my_file = pyfits.open(filename)
```
I don't know if this was your choice or not but I am *strong* :-1: on `pyfits`, the recommended way from astropy and sunpy is:

```python
from astropy.io import fits

fits.open(....
```

BaptistePellorceAstro replied:
I know but on the sji.py file, pyfits is used : this is not my choice and I don't want to change this part. However, for my next pull request, I can modify it if you want (I also prefer fits than pyfits !)

Cadair replied:
If you want to change it in a separate PR that sounds awesome!

DanRyanIrish replied:
In fact, ```astropy.io.fits``` is already being used.  See the import statement which says:
```from astropy.io import fits as pyfits``` (https://github.com/sunpy/irispy/pull/67/files#diff-c3a5b1ac9cb9013bd70b2db039f006b3R15)
@BaptistePellorceAstro, could you delete the  ```as pyfits``` part of that line and then to a search and replace changing ```pyfits``` to ```fits```?

BaptistePellorceAstro replied:
Yes, I planned to do it.

DanRyanIrish replied:
👍

Cadair commented on irispy/sji.py:
```diff
@@ -878,3 +879,73 @@ def dustbuster(mc):
+    # Extraction of meta for NDCube from fits file.
+    try:
+        date_obs = parse_time(my_file[0].header["DATE_OBS"])
+    except ValueError:
```
rather than using try/except for doing this you can use `dict.get` i.e:

```python
date_obs = myfile[0].header.get('DATE_OBS', None)
```
You will then need to only send to `parse_time`, if it's not `None`
```python
date_obs = parse_time(date_obs) if date_obs else None
```

(As a side note, it's a good idea to avoid using try/except blocks if there is a simple way to do so. They are quite expensive and also can lead to harder to debug errors.)

DanRyanIrish replied:
I didn't realise ```try...except...``` statements were so expensive.  That's really good to know.  I wrote it in this way for the spectrograph fits reader function.  So I should probably rewrite that in the future along the lines you suggested.  @BaptistePellorceAstro, the SJI reader should probably also be rewritten as @Cadair said.  Looks like we both have homework ;)

BaptistePellorceAstro replied:
I planned doing it too

DanRyanIrish replied:
👍

BaptistePellorceAstro commented:
Thanks for your returns, they will take part of my next pull request. 

Do you have some ideas about `SJICube` refactor that I can now work about ?

Cadair commented:
@BaptistePellorceAstro I don't know how familiar with GitHub you are, but you can make changes to this PR by pushing more changes to your fork. (On the same `master` branch).

DanRyanIrish commented:
Hi @BaptistePellorceAstro.  Thanks so much for your first PR (pull request) to IRISpy!  The ```SJICube``` refactor will be a very useful addition to IRISpy.  Now starts the valuable, although sometimes frustrating, process of code review.  The comments and suggestions are meant maximise the quality and usefulness of your PR, and in doing so make you a better Python programmer.  They are never meant to criticise or discourage.  My first ever PR has gone down in SunPy legend as the longest and most frustrating code review process ever!  However, in the end it gave me the Python skills and confidence to eventually create IRISpy.  So the process worked!  We all get the same treatment and I think there's very little chance of your PR experience being as bad as my first one ;)

As we make suggestions and request changes, you can make those changes to the same branch that you've PR'd (in this case your local master branch) and this ```git push``` the new commits to the branch on your github account.  As @Cadair said, the changes will automatically appear as part of this PR.  Comments made on lines of code you've changed will be marked as outdated, while comments on unchanged lines remain as they were.  This way we go through an iterative process in getting your PR to a point where it's ready to merge into IRISpy and be used by everyone!

I'll leave some specific comments on the code soon.

DanRyanIrish commented:
@BaptistePellorceAstro regarding your query about the ```SJICube``` class, let's get the FITS reader right first.

BaptistePellorceAstro commented:
Thanks for your answer, I hope I will have fun to work on IRISpy. As I said previously, I am open to every suggestion by the fact I am learning python ! I don't have access to my working computer now, I will modify it tomorrow morning.

Thanks for all returns and teach me python !

DanRyanIrish commented on irispy/sji.py:
```diff
@@ -878,3 +879,73 @@ def dustbuster(mc):
+    meta = {'TELESCOP': my_file[0].header.get('TELESCOP'),
+            'INSTRUME': my_file[0].header.get('INSTRUME'),
+            'TWAVE1': my_file[0].header.get('TWAVE1'),
+            'WAVEUNIT': "Angstrom",
```
@BaptistePellorceAstro I think this PR isn't looking too bad so far :) One question, did you just copy all the metadata stuff from the spectrograph reader or did you go through a SJI FITS file and only include the metadata from that?  For example, does this metadata (```'WAVEUNIT'```) exist in SJI files?

BaptistePellorceAstro replied:
I found "WAVEUNIT" on the SJICube.__init__ and I was thinking that will be more clear to read it.
I will check if I can found it in FITS files.

Initially, my function was calling a new class that I deleted to input reading function in sji.py
This meta data is a remanant of this call.

DanRyanIrish commented:
You're welcome @BaptistePellorceAstro.  I found code review an intense but great way to learn a language.  Thanks to you for contributing!

BaptistePellorceAstro commented:
You can find a new version of the FITS reading function. 

1) I changed `from astropy.io import fits as pyfits` to `from astropy.io import fits` and the two `pyfits` calls to `fits`.

2) I deleted `try ... except ...` and changed it by @Cadair suggestion.

About metadata, I pick up some ideas from spectrograph.py and sji.py and I mixed them. `'WAVEUNIT'` is not existing on SJI FITS files but I think it will be useful to have a better understanding of metadata. However, I can remove it if you want.

BaptistePellorceAstro commented:
Sorry for this two successive commits but deleting `as pyfits` created an error on `SJICube._init_` because the name of the fits file was `fits`

DanRyanIrish commented on irispy/sji.py:
```diff
@@ -878,3 +879,69 @@ def dustbuster(mc):
+    meta = {'TELESCOP': my_file[0].header.get('TELESCOP'),
+            'INSTRUME': my_file[0].header.get('INSTRUME'),
+            'TWAVE1': my_file[0].header.get('TWAVE1'),
+            'WAVEUNIT': "Angstrom",
```
```SJICube``` (which the output of this function will become) should not contain any spectrograph-specific metadata.  So I think it's best to remove ```'WAVEUNIT'```.  Same goes for any other spectrograph-specific metadata.

BaptistePellorceAstro replied:
Ok, I will remove it.

DanRyanIrish commented on irispy/sji.py:
```diff
@@ -878,3 +879,69 @@ def dustbuster(mc):
+    xcenix = my_file[1].data[:, my_file[1].header["XCENIX"]] * u.arcsec
+    ycenix = my_file[1].data[:, my_file[1].header["YCENIX"]] * u.arcsec
+    obs_vrix = my_file[1].data[:, my_file[1].header["OBS_VRIX"]] * u.m/u.s
+    ophaseix = my_file[1].data[:, my_file[1].header["OPHASEIX"]]
```
Have you confirmed that all these properties are in SJI FITS files?

BaptistePellorceAstro replied:
I am not at my office but I saw on tutorial that SJI headers have these properties. For now, I can say yes but I will try to go to office next week (hollidays) to check and to be 100 percent sure of that.

BaptistePellorceAstro replied:
Sorry to have made you wait last week. I certify that these 4 datas are in the my_file[1].header with the same name.

DanRyanIrish commented:
Well apartment from the metadata issue I think this PR is looking good!

DanRyanIrish commented:
Hi @BaptistePellorceAstro.  How is this coming along?

DanRyanIrish commented:
Hi @BaptistePellorceAstro.  No problem.  So do you feel this PR is ready for final code review or do you think you have more to do?

BaptistePellorceAstro commented:
I think we can do this PR, I finished to worked on the reading function.

DanRyanIrish commented:
I'm happy with these changes 👍 If we do find bugs though you can fix them in a new PR.
